In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import StandardScaler

# Configure numbers so easier to read
pd.options.display.float_format = '{:20.2f}'.format

# Show all columns on output
pd.set_option('display.max_columns', 999)

# Data Exploration

First, let's load in the first sheet of data as a dataframe (2009-2010) to see what we are working with.

In [2]:
df = pd.read_excel('./data/online_retail_II.xlsx', sheet_name=0)
df.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate                Price          Customer ID  \
0 2009-12-01 07:45:00                 6.95             13085.00   
1 2009-12-01 07:45:00                 6.75             13085.00   
2 2009-12-01 07:45:00                 6.75             13085.00   
3 2009-12-01 07:45:00                 2.10             13085.00   
4 2009-12-01 07:45:00                 1.25             13085.00   
5 2009-12-01 07:45:00                 1.65             13085.00   
6 2009-12-01 07:45:00                 1.25             13085.00   
7 2009-12-01 07:45:00                 5.95             13085.00   
8 2009-12-01 07:46:00                 2.55             13085.00   
9 2009-12-01 07:46:00                 3.75             13085.00   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom  
5  United Kingdom  
6  United Kingdom  
7  United Kingdom  
8  United Kingdom  
9  United Kingdom

Next, we can learn more about the type of data we are working with and what type of format they are in.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


Looks good! All our string columns are represented as `objects`, the `InvoiceDate` is automatically detected as a `datetime` which is awesome, and our numeric fields seem okay too. It's odd to see a `Customer ID` column as a float (I would expect an integer instead!), we'll have to take a look at that later.

Okay, let's now look at the basic statistics for the quantitative data.

In [4]:
df.describe()

Quantity                    InvoiceDate  \
count            525461.00                         525461   
mean                 10.34  2010-06-28 11:37:36.845017856   
min               -9600.00            2009-12-01 07:45:00   
25%                   1.00            2010-03-21 12:20:00   
50%                   3.00            2010-07-06 09:51:00   
75%                  10.00            2010-10-15 12:45:00   
max               19152.00            2010-12-09 20:01:00   
std                 107.42                            NaN   

                     Price          Customer ID  
count            525461.00            417534.00  
mean                  4.69             15360.65  
min              -53594.36             12346.00  
25%                   1.25             13983.00  
50%                   2.10             15311.00  
75%                   4.21             16799.00  
max               25111.09             18287.00  
std                 146.13              1680.81

Now that's interesting, a minus value as the minimum quantity and price? Suspicious. The maximum `price` seems a little high too. This is sales data minus values shouldn't exist, surely? Something can't be right there. We'll have to investigate that.

Now let's check out the qualitative data.

In [5]:
df.describe(include=[object])

Invoice StockCode                         Description         Country
count    525461    525461                              522533          525461
unique    28816      4632                                4681              40
top      537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq        675      3516                                3549          485852

Hmmm... there looks to be some missing values in the `Description` column and a minor disparity between the frequencies. We'll take a look at that.

## Missing Values & Oddities

Alright, next let's take a look at our missing data.

In [6]:
df.isna().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [7]:
df.isna().sum().sum()

np.int64(110855)

It looks like only two of our columns have missing data with a combined total of `110,855` missing cells. `Customer ID` is our biggest concern so that's check that out first.

In [8]:
df[df["Customer ID"].isna()].head(10)

Invoice StockCode                   Description  Quantity  \
263   489464     21733                  85123a mixed       -96   
283   489463     71477                         short      -240   
284   489467    85123A                   21733 mixed      -192   
470   489521     21646                           NaN       -50   
577   489525    85226C     BLUE PULL BACK RACING CAR         1   
578   489525     85227   SET/6 3D KIT CARDS FOR KIDS         1   
1055  489548     22271          FELTCRAFT DOLL ROSIE         1   
1056  489548     22254          FELT TOADSTOOL LARGE        12   
1057  489548     22273          FELTCRAFT DOLL MOLLY         3   
1058  489548     22195  LARGE HEART MEASURING SPOONS         1   

             InvoiceDate                Price  Customer ID         Country  
263  2009-12-01 10:52:00                 0.00          NaN  United Kingdom  
283  2009-12-01 10:52:00                 0.00          NaN  United Kingdom  
284  2009-12-01 10:53:00                 0.00          NaN  United Kingdom  
470  2009-12-01 11:44:00                 0.00          NaN  United Kingdom  
577  2009-12-01 11:49:00                 0.55          NaN  United Kingdom  
578  2009-12-01 11:49:00                 0.85          NaN  United Kingdom  
1055 2009-12-01 12:32:00                 2.95          NaN  United Kingdom  
1056 2009-12-01 12:32:00                 1.25          NaN  United Kingdom  
1057 2009-12-01 12:32:00                 2.95          NaN  United Kingdom  
1058 2009-12-01 12:32:00                 1.65          NaN  United Kingdom

This data has some weird behaviour going on. Minus quanities, with 0 pricing - could these perhaps be thefts or stock write-offs? It's hard to tell.

Hmmm.. We'll that's annoying. Invoices 577 onwards in our sample seem to be valid transactions, but without a `Customer ID` we cannot use them, there's no way to infer which transaction the customer relates to. Our only option here is to drop these missing values from our main analysis.

Before we drop them though, let's take a look at these negative quantities.

In [9]:
df[df["Quantity"] < 0].head(10)

Invoice StockCode                        Description  Quantity  \
178  C489449     22087           PAPER BUNTING WHITE LACE       -12   
179  C489449    85206A       CREAM FELT EASTER EGG BASKET        -6   
180  C489449     21895      POTTING SHED SOW 'N' GROW SET        -4   
181  C489449     21896                 POTTING SHED TWINE        -6   
182  C489449     22083         PAPER CHAIN KIT RETRO SPOT       -12   
183  C489449     21871                SAVE THE PLANET MUG       -12   
184  C489449     84946    ANTIQUE SILVER TEA GLASS ETCHED       -12   
185  C489449    84970S  HANGING HEART ZINC T-LIGHT HOLDER       -24   
186  C489449     22090          PAPER BUNTING RETRO SPOTS       -12   
196  C489459    90200A         PURPLE SWEETHEART BRACELET        -3   

            InvoiceDate                Price          Customer ID  \
178 2009-12-01 10:33:00                 2.95             16321.00   
179 2009-12-01 10:33:00                 1.65             16321.00   
180 2009-12-01 10:33:00                 4.25             16321.00   
181 2009-12-01 10:33:00                 2.10             16321.00   
182 2009-12-01 10:33:00                 2.95             16321.00   
183 2009-12-01 10:33:00                 1.25             16321.00   
184 2009-12-01 10:33:00                 1.25             16321.00   
185 2009-12-01 10:33:00                 0.85             16321.00   
186 2009-12-01 10:33:00                 2.95             16321.00   
196 2009-12-01 10:44:00                 4.25             17592.00   

            Country  
178       Australia  
179       Australia  
180       Australia  
181       Australia  
182       Australia  
183       Australia  
184       Australia  
185       Australia  
186       Australia  
196  United Kingdom

Okay, so a lot of these invoices look like cancellations (invoices starting with a `C` means cancellation). That's expected based on what we know about our initial understanding. Let's see how many records we are looking at.

In [10]:
negative_quantites = df[df["Quantity"] < 0]
negative_quantites.shape

(12326, 8)

In [11]:
negative_quantites["Invoice"].str.startswith('C').sum(), negative_quantites["Invoice"].str.startswith('C').isna().sum()

(10205, np.int64(2121))

Out of all of them (less than 0 prices), about 20% relate to missing records. Let's explore things further and see what other types of invoice patterns we have. We should only have two formats: `6 numbers`, `C + 6 numbers`.

Firstly, let's change the `Invoice` column to a string so we can use REGEX. Next, we'll 'remove' the digits and see what letters we have left.

In [12]:
df["Invoice"] = df["Invoice"].astype("str")

In [13]:
df["Invoice"].str.replace("[0-9]", "", regex=True).unique()

array(['', 'C', 'A'], dtype=object)

Well that's unexpected! What is this `A` all about? Let's take a look.

In [14]:
df[df["Invoice"].str.startswith("A")]

Invoice StockCode      Description  Quantity         InvoiceDate  \
179403  A506401         B  Adjust bad debt         1 2010-04-29 13:36:00   
276274  A516228         B  Adjust bad debt         1 2010-07-19 11:24:00   
403472  A528059         B  Adjust bad debt         1 2010-10-20 12:04:00   

                      Price  Customer ID         Country  
179403            -53594.36          NaN  United Kingdom  
276274            -44031.79          NaN  United Kingdom  
403472            -38925.87          NaN  United Kingdom

These look like pure accounting transactions and hold no real value in analysing our customers behaviour. The large negative price values would also skew our results drastically. We'll remove these from our analysis. 

Now that we've seen some issues with the `InvoiceNo`, let's check out the `StockCode` column and make sure that its in the correct format too. This should only consist of `5` digit numeric codes.

In [15]:
df.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate                Price          Customer ID  \
0 2009-12-01 07:45:00                 6.95             13085.00   
1 2009-12-01 07:45:00                 6.75             13085.00   
2 2009-12-01 07:45:00                 6.75             13085.00   
3 2009-12-01 07:45:00                 2.10             13085.00   
4 2009-12-01 07:45:00                 1.25             13085.00   
5 2009-12-01 07:45:00                 1.65             13085.00   
6 2009-12-01 07:45:00                 1.25             13085.00   
7 2009-12-01 07:45:00                 5.95             13085.00   
8 2009-12-01 07:46:00                 2.55             13085.00   
9 2009-12-01 07:46:00                 3.75             13085.00   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom  
5  United Kingdom  
6  United Kingdom  
7  United Kingdom  
8  United Kingdom  
9  United Kingdom

Just by taking another quick look at some of the records, we can immediately see some issues with the `StockCode`. Let's do a deeper dive into it. Using the same tactic as before by 'removing' our numbers. 

In [16]:
df["StockCode"] = df["StockCode"].astype("str")

In [17]:
df["StockCode"].str.replace("[0-9]", "", regex=True).unique()

array(['', 'P', 'W', 'C', 'B', 'F', 'L', 'S', 'A', 'N', 'POST', 'E', 'J',
       'D', 'G', 'LP', 'BL', 'K', 'H', 'GR', 'M', 'DCGS', 'DOT', 'U', 'b',
       'w', 'c', 'a', 'f', 'bl', 's', 'p', 'R', 'V', 'T', 'I',
       'BANK CHARGES', 'O', 'Z', 'TEST', 'gift__', 'DCGSN', 'm', 'PADS',
       'Y', 'HC', 'e', 'd', 'ADJUST', 'DCGSSGIRL', 'GIFT', 'DCGSLBOY',
       'k', 'g', 'DCGSSBOY', 'DCGSLGIRL', 'j', 'l', 'n', 'J ', 'SP',
       'AMAZONFEE'], dtype=object)

In [18]:
df[(df["StockCode"].str.match('^\\d{5}$') == False)]

Invoice StockCode                          Description  Quantity  \
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
12      489436    48173C                DOOR MAT BLACK FLOCK         10   
23      489436    35004B          SET OF 3 BLACK FLYING DUCKS        12   
28      489436    84596F         SMALL MARSHMALLOWS PINK BOWL         8   
...        ...       ...                                  ...       ...   
525387  538170    84029E       RED WOOLLY HOTTIE WHITE HEART.         2   
525388  538170    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         2   
525389  538170    85232B      SET OF 3 BABUSHKA STACKING TINS         2   
525435  538171    47591D      PINK FAIRY CAKE CHILDRENS APRON         1   
525436  538171    47591B             SCOTTIES CHILDRENS APRON         2   

               InvoiceDate                Price          Customer ID  \
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
12     2009-12-01 09:06:00                 5.95             13078.00   
23     2009-12-01 09:06:00                 4.65             13078.00   
28     2009-12-01 09:06:00                 1.25             13078.00   
...                    ...                  ...                  ...   
525387 2010-12-09 19:32:00                 3.75             13969.00   
525388 2010-12-09 19:32:00                 3.75             13969.00   
525389 2010-12-09 19:32:00                 4.95             13969.00   
525435 2010-12-09 20:01:00                 1.95             17530.00   
525436 2010-12-09 20:01:00                 1.65             17530.00   

               Country  
1       United Kingdom  
2       United Kingdom  
12      United Kingdom  
23      United Kingdom  
28      United Kingdom  
...                ...  
525387  United Kingdom  
525388  United Kingdom  
525389  United Kingdom  
525435  United Kingdom  
525436  United Kingdom  

[80112 rows x 8 columns]

Here we can see that numbers are not the only item in the `StockCode` column. Some of them look like specific sale events, you also have test records and bank charges, which can be removed. The single letters look like valid transactions so we will keep them in our analysis. Let's now check the other values, such as `POST`, `DOT`, `PADS`, etc.

Regex notes:
- `^\\d{5}$` - 5 digits only (`[0-9]`), this is what we expect our `StockCode` values to be
- `^\\d{5}[a-zA-Z]+$` - any values starting with 5 digits and ending in text

In [19]:
odd_stockcode_data = df[(df["StockCode"].str.match('^\\d{5}$') == False) & (df["StockCode"].str.match('^\\d{5}[a-zA-Z]+$') == False)]
odd_stockcode_data

Invoice StockCode     Description  Quantity         InvoiceDate  \
89       489439      POST         POSTAGE         3 2009-12-01 09:28:00   
126      489444      POST         POSTAGE         1 2009-12-01 09:55:00   
173      489447      POST         POSTAGE         1 2009-12-01 10:10:00   
625      489526      POST         POSTAGE         6 2009-12-01 11:50:00   
735     C489535         D        Discount        -1 2009-12-01 12:11:00   
...         ...       ...             ...       ...                 ...   
524776   538147         M          Manual         1 2010-12-09 16:11:00   
524887   538148       DOT  DOTCOM POSTAGE         1 2010-12-09 16:26:00   
525000   538149       DOT  DOTCOM POSTAGE         1 2010-12-09 16:27:00   
525126   538153       DOT  DOTCOM POSTAGE         1 2010-12-09 16:31:00   
525147   538154       DOT  DOTCOM POSTAGE         1 2010-12-09 16:35:00   

                      Price          Customer ID         Country  
89                    18.00             12682.00          France  
126                  141.00             12636.00             USA  
173                  130.00             12362.00         Belgium  
625                   18.00             12533.00         Germany  
735                    9.00             15299.00  United Kingdom  
...                     ...                  ...             ...  
524776                15.00             13090.00  United Kingdom  
524887               547.32                  NaN  United Kingdom  
525000               620.68                  NaN  United Kingdom  
525126               822.94                  NaN  United Kingdom  
525147                85.79                  NaN  United Kingdom  

[3098 rows x 8 columns]

In [20]:
odd_stockcode_data["StockCode"].unique()

array(['POST', 'D', 'DCGS0058', 'DCGS0068', 'DOT', 'M', 'DCGS0004',
       'DCGS0076', 'C2', 'BANK CHARGES', 'DCGS0003', 'TEST001',
       'gift_0001_80', 'DCGS0072', 'gift_0001_20', 'DCGS0044', 'TEST002',
       'gift_0001_10', 'gift_0001_50', 'DCGS0066N', 'gift_0001_30',
       'PADS', 'ADJUST', 'gift_0001_40', 'gift_0001_60', 'gift_0001_70',
       'gift_0001_90', 'DCGSSGIRL', 'DCGS0006', 'DCGS0016', 'DCGS0027',
       'DCGS0036', 'DCGS0039', 'DCGS0060', 'DCGS0056', 'DCGS0059', 'GIFT',
       'DCGSLBOY', 'm', 'DCGS0053', 'DCGS0062', 'DCGS0037', 'DCGSSBOY',
       'DCGSLGIRL', 'S', 'DCGS0069', 'DCGS0070', 'DCGS0075', 'B',
       'DCGS0041', 'ADJUST2', '47503J ', 'C3', 'SP1002', 'AMAZONFEE'],
      dtype=object)

To keep things simple, I've analysed all of the above unique codes in the cell below (replacing the code to prevent the notebook from getting too long). Here's what we've learned:

| Code              | Description                                                   | Exclude? | 
|-------------------|---------------------------------------------------------------|----------| 
| POST              | Looks valid, postage charges                                  | &check;  |
| D                 | Looks valid, discount values                                  | &check;  |
| DCGSXXXX          | Looks valid, some quantities are negative, customer ID null   | &check;  |
| DOT               | Looks valid, dotcom postage charges                           | &check;  |
| M, m              | Looks valid, manual transactions                              | &check;  |
| C2                | Not sure, Carriage transaction                                | &check;  |
| BANK CHARGES, B   | Bank charges                                                  | &check;  |
| TESTXXX           | Test data                                                     | &check;  |
| GIFT, gift__XXX   | Gift card purchases, another analysis(?), customer ID null    | &check;  |
| PADS              | Looks valid, pad related projects                             | &cross;  |
| ADJUSTX           | Manual account adjustments by admins                          | &check;  |
| C3                | Not sure, only 1 transaction                                  | &check;  |
| S                 | Samples sent to customer                                      | &check;  |
| SP1002            | Special request items, look valid, 3 items, 1 Cutomer ID null | &check;  |
| AMAZONFEE         | Amazon shipping fees, customer ID null                        | &check;  |

We'll keep the `PADS` but remove everything else. That about does it for the `StockCode` column. Onto data cleaning!

In [21]:
odd_stockcode_data[odd_stockcode_data["StockCode"].str.contains("AMAZONFEE")]

Invoice  StockCode Description  Quantity         InvoiceDate  \
440688  C531400  AMAZONFEE  AMAZON FEE        -1 2010-11-08 10:08:00   
440698   531411  AMAZONFEE  AMAZON FEE         1 2010-11-08 10:11:00   
517452  C537600  AMAZONFEE  AMAZON FEE        -1 2010-12-07 12:41:00   
517953  C537630  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:04:00   
517955   537632  AMAZONFEE  AMAZON FEE         1 2010-12-07 15:08:00   
519170  C537644  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:34:00   
519251  C537647  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:41:00   
519294  C537651  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:49:00   
519295  C537652  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:51:00   

                      Price  Customer ID         Country  
440688              6706.71          NaN  United Kingdom  
440698              6706.71          NaN  United Kingdom  
517452                 1.00          NaN  United Kingdom  
517953             13541.33          NaN  United Kingdom  
517955             13541.33          NaN  United Kingdom  
519170             13474.79          NaN  United Kingdom  
519251              5519.25          NaN  United Kingdom  
519294             13541.33          NaN  United Kingdom  
519295              6706.71          NaN  United Kingdom

### Notes For Data Cleaning

Row removal -

- Has the following `StockCodes`:

    `['POST', 'D', 'DCGS0058', 'DCGS0068', 'DOT', 'M', 'DCGS0004', 'DCGS0076', 'C2', 'BANK CHARGES', 'DCGS0003', 'TEST001', 'gift_0001_80', 'DCGS0072', 'gift_0001_20', 'DCGS0044', 'TEST002', 'gift_0001_10', 'gift_0001_50', 'DCGS0066N', 'gift_0001_30', 'ADJUST', 'gift_0001_40', 'gift_0001_60', 'gift_0001_70', 'gift_0001_90', 'DCGSSGIRL', 'DCGS0006', 'DCGS0016', 'DCGS0027', 'DCGS0036', 'DCGS0039', 'DCGS0060', 'DCGS0056', 'DCGS0059', 'GIFT', 'DCGSLBOY', 'm', 'DCGS0053', 'DCGS0062', 'DCGS0037', 'DCGSSBOY', 'DCGSLGIRL', 'S', 'DCGS0069', 'DCGS0070', 'DCGS0075', 'B', 'DCGS0041', 'ADJUST2', '47503J ', 'C3', 'SP1002', 'AMAZONFEE']`

- Has a missing `Customer ID`
- `InvoiceNo` starts with `A` or `C`



## Data Cleaning

Now let's move onto cleaning are data based on the above requirements. We start by making a copy of our original data before cleaning.

In [22]:
cleaned_df = df.copy()
cleaned_df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[525461 rows x 8 columns]

We then create a mask of our `InvoiceNo` column that performs the required updates.  

In [23]:
invoice_mask = cleaned_df["Invoice"].str.match("^\\d{6}$") == True

And, do the same for the `StockCode` column then apply the masks to the cleaned DataFrame.

In [24]:
stockcode_mask = (
    (cleaned_df["StockCode"].str.match('^\\d{5}$') == True)  # 5 digits
    | (cleaned_df["StockCode"].str.match('^\\d{5}[a-zA-Z]+$') == True)  # 5 digits with text after it
    | (cleaned_df["StockCode"].str.match('^PADS$') == True) # PADS data
)

In [25]:
cleaned_df = cleaned_df[invoice_mask]
cleaned_df = cleaned_df[stockcode_mask]
cleaned_df

C:\Users\rpart\AppData\Local\Temp\ipykernel_20412\4045565412.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  cleaned_df = cleaned_df[stockcode_mask]


Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[512796 rows x 8 columns]

Now we remove the missing `Customer ID` rows.

In [26]:
cleaned_df.dropna(subset=["Customer ID"], inplace=True)

In [27]:
cleaned_df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[406337 rows x 8 columns]

Perfect! Next, let's take a look at our column stats to see our changes.

In [28]:
cleaned_df.describe()

Quantity                    InvoiceDate  \
count            406337.00                         406337   
mean                 13.62  2010-07-01 10:11:06.543288320   
min                   1.00            2009-12-01 07:45:00   
25%                   2.00            2010-03-26 14:01:00   
50%                   5.00            2010-07-09 15:48:00   
75%                  12.00            2010-10-14 17:09:00   
max               19152.00            2010-12-09 20:01:00   
std                  97.00                            NaN   

                     Price          Customer ID  
count            406337.00            406337.00  
mean                  2.99             15373.63  
min                   0.00             12346.00  
25%                   1.25             14004.00  
50%                   1.95             15326.00  
75%                   3.75             16814.00  
max                 295.00             18287.00  
std                   4.29              1677.37

In [29]:
cleaned_df.describe(include=[object])

Invoice StockCode                         Description         Country
count   406337    406337                              406337          406337
unique   18966      4007                                4433              37
top     500356    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq       270      3153                                3153          370487

Awesome! The minimum `quantity` and `price` are now back to normal and everything else looks great. Although, a `price` of 0 is a little odd. There shouldn't be any freebies! Let's take a look.

In [30]:
cleaned_df[cleaned_df["Price"] == 0]

Invoice StockCode                        Description  Quantity  \
4674    489825     22076                 6 RIBBONS EMPIRE          12   
6781    489998     48185                DOOR MAT FAIRY CAKE         2   
18738   490961     22065     CHRISTMAS PUDDING TRINKET POT          1   
18739   490961     22142       CHRISTMAS CRAFT WHITE FAIRY         12   
32916   492079     85042          ANTIQUE LILY FAIRY LIGHTS         8   
40101   492760     21143    ANTIQUE GLASS HEART DECORATION         12   
47126   493761     79320                    FLAMINGO LIGHTS        24   
48342   493899     22355        CHARLOTTE BAG , SUKI DESIGN        10   
57619   494607     21533          RETRO SPOT LARGE MILK JUG        12   
111348  500073     21662         VINTAGE GLASS COFFEE CADDY         1   
149201  503585     22459       CAST IRON HOOK GARDEN TROWEL         8   
149202  503585     22458         CAST IRON HOOK GARDEN FORK         8   
166143  505083     22376  AIRLINE BAG VINTAGE JET SET WHITE         1   
232526  511902     21765            HANGING METAL BIRD BATH         1   
240455  512609     20914   SET/5 RED SPOTTY LID GLASS BOWLS         2   
248583  513416     22423           REGENCY CAKESTAND 3 TIER         5   
276858  516304     22690      DOORMAT HOME SWEET HOME BLUE          6   
296375  518231     22472          TV DINNER TRAY DOLLY GIRL         9   
327801  521375     22202            MILK PAN PINK RETROSPOT         3   
358820  524181    46000M       POLYESTER FILLER PAD 45x45cm       648   
364333  524701     22218              CAKE STAND LACE WHITE         2   
392008  527084     22630               DOLLY GIRL LUNCH BOX        64   
400047  527696     22121         NOEL WOODEN BLOCK LETTERS          1   
439309  531361     21843           RED RETROSPOT CAKE STAND         2   
453705  532470     22624               IVORY KITCHEN SCALES         2   
471775  533822     22846         BREAD BIN DINER STYLE RED          1   
471776  533822     22845   VINTAGE CREAM CAT FOOD CONTAINER         1   
512240  537197     22841       ROUND CAKE TIN VINTAGE GREEN         1   

               InvoiceDate                Price          Customer ID  \
4674   2009-12-02 13:34:00                 0.00             16126.00   
6781   2009-12-03 11:19:00                 0.00             15658.00   
18738  2009-12-08 15:25:00                 0.00             14108.00   
18739  2009-12-08 15:25:00                 0.00             14108.00   
32916  2009-12-15 13:49:00                 0.00             15070.00   
40101  2009-12-18 14:22:00                 0.00             18071.00   
47126  2010-01-06 14:54:00                 0.00             14258.00   
48342  2010-01-08 10:43:00                 0.00             12417.00   
57619  2010-01-15 12:43:00                 0.00             16858.00   
111348 2010-03-04 11:44:00                 0.00             13047.00   
149201 2010-04-01 17:13:00                 0.00             13047.00   
149202 2010-04-01 17:13:00                 0.00             13047.00   
166143 2010-04-20 09:56:00                 0.00             12623.00   
232526 2010-06-11 11:12:00                 0.00             12748.00   
240455 2010-06-17 10:12:00                 0.00             14045.00   
248583 2010-06-24 12:34:00                 0.00             13089.00   
276858 2010-07-19 13:13:00                 0.00             14025.00   
296375 2010-08-05 15:28:00                 0.00             12471.00   
327801 2010-09-05 11:58:00                 0.00             12647.00   
358820 2010-09-27 16:59:00                 0.00             17450.00   
364333 2010-09-30 12:19:00                 0.00             17667.00   
392008 2010-10-14 15:33:00                 0.00             14646.00   
400047 2010-10-18 15:13:00                 0.00             13554.00   
439309 2010-11-07 14:26:00                 0.00             12820.00   
453705 2010-11-12 11:41:00                 0.00             12647.00   
471775 2010-11-19

These are definitely red flags. 648 items for free?! I'm not sure why these are here, I'd be talking to management about these ones! Let's remove them and check the stats again.

In [31]:
cleaned_df = cleaned_df[cleaned_df["Price"] > 0]

In [32]:
cleaned_df.describe()

Quantity                    InvoiceDate  \
count            406309.00                         406309   
mean                 13.62  2010-07-01 10:14:25.869572352   
min                   1.00            2009-12-01 07:45:00   
25%                   2.00            2010-03-26 14:01:00   
50%                   5.00            2010-07-09 15:48:00   
75%                  12.00            2010-10-14 17:09:00   
max               19152.00            2010-12-09 20:01:00   
std                  97.00                            NaN   

                     Price          Customer ID  
count            406309.00            406309.00  
mean                  2.99             15373.72  
min                   0.00             12346.00  
25%                   1.25             14006.00  
50%                   1.95             15326.00  
75%                   3.75             16814.00  
max                 295.00             18287.00  
std                   4.29              1677.33

Er... the price appears to still be `0`. Let's just quickly check that and make sure it is larger than it.

In [33]:
cleaned_df["Price"].min()

np.float64(0.001)

In [34]:
cleaned_df[(cleaned_df["Price"] > 0) & (cleaned_df["Price"] < 1)]["Price"].unique()

array([0.85 , 0.65 , 0.98 , 0.42 , 0.72 , 0.55 , 0.38 , 0.21 , 0.75 ,
       0.95 , 0.29 , 0.34 , 0.19 , 0.64 , 0.37 , 0.1  , 0.14 , 0.36 ,
       0.18 , 0.12 , 0.08 , 0.06 , 0.5  , 0.25 , 0.48 , 0.32 , 0.67 ,
       0.22 , 0.16 , 0.04 , 0.79 , 0.09 , 0.53 , 0.17 , 0.7  , 0.001,
       0.4  , 0.3  , 0.15 , 0.88 , 0.63 , 0.24 , 0.11 , 0.45 , 0.35 ,
       0.8  , 0.6  , 0.52 , 0.2  , 0.66 , 0.07 , 0.87 , 0.62 , 0.78 ,
       0.27 , 0.13 , 0.31 , 0.33 , 0.05 , 0.03 , 0.89 , 0.74 , 0.9  ,
       0.73 ])

Okay, so it looks like the smallest item value sold is `0.001` which doesn't seem right. That's a 10th of a penny!

In [35]:
print(cleaned_df[cleaned_df["Price"] == 0.001].shape[0])
cleaned_df[cleaned_df["Price"] == 0.001]

14


Invoice StockCode                 Description  Quantity  \
62299   494914      PADS  PADS TO MATCH ALL CUSHIONS         1   
74731   496222      PADS  PADS TO MATCH ALL CUSHIONS         1   
77702   496473      PADS  PADS TO MATCH ALL CUSHIONS         1   
79794   496643      PADS  PADS TO MATCH ALL CUSHIONS         1   
90798   497935      PADS  PADS TO MATCH ALL CUSHIONS         1   
97716   498562      PADS  PADS TO MATCH ALL CUSHIONS         1   
101718  499056      PADS  PADS TO MATCH ALL CUSHIONS         1   
104480  499399      PADS  PADS TO MATCH ALL CUSHIONS         1   
123947  501176      PADS  PADS TO MATCH ALL CUSHIONS         1   
156809  504332      PADS  PADS TO MATCH ALL CUSHIONS         1   
178629  506303      PADS  PADS TO MATCH ALL CUSHIONS         1   
341826  522766      PADS  PADS TO MATCH ALL CUSHIONS         1   
440930  531433      PADS  PADS TO MATCH ALL CUSHIONS         1   
468280  533528      PADS  PADS TO MATCH ALL CUSHIONS         1   

               InvoiceDate                Price          Customer ID  \
62299  2010-01-19 17:04:00                 0.00             16705.00   
74731  2010-01-29 13:53:00                 0.00             13583.00   
77702  2010-02-01 15:38:00                 0.00             17350.00   
79794  2010-02-03 11:58:00                 0.00             13408.00   
90798  2010-02-15 10:47:00                 0.00             13408.00   
97716  2010-02-21 12:03:00                 0.00             15182.00   
101718 2010-02-24 13:46:00                 0.00             13765.00   
104480 2010-02-26 13:26:00                 0.00             14459.00   
123947 2010-03-15 11:00:00                 0.00             14857.00   
156809 2010-04-12 16:30:00                 0.00             12671.00   
178629 2010-04-29 10:43:00                 0.00             14249.00   
341826 2010-09-16 13:58:00                 0.00             15182.00   
440930 2010-11-08 11:54:00                 0.00             13317.00   
468280 2010-11-17 15:08:00                 0.00             15750.00   

               Country  
62299   United Kingdom  
74731   United Kingdom  
77702   United Kingdom  
79794   United Kingdom  
90798   United Kingdom  
97716   United Kingdom  
101718  United Kingdom  
104480  United Kingdom  
123947  United Kingdom  
156809         Germany  
178629  United Kingdom  
341826  United Kingdom  
440930  United Kingdom  
468280  United Kingdom

It looks like they are cushion related. I guess nothing can be free with this company? We'll remove them from our analysis too.

In [36]:
cleaned_df = cleaned_df[cleaned_df["Price"] != 0.001]
cleaned_df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
525456 2010-12-09 20:01:00                 2.95             17530.00   
525457 2010-12-09 20:01:00                 3.75             17530.00   
525458 2010-12-09 20:01:00                 3.75             17530.00   
525459 2010-12-09 20:01:00                 3.75             17530.00   
525460 2010-12-09 20:01:00                 1.95             17530.00   

               Country  
0       United Kingdom  
1       United Kingdom  
2       United Kingdom  
3       United Kingdom  
4       United Kingdom  
...                ...  
525456  United Kingdom  
525457  United Kingdom  
525458  United Kingdom  
525459  United Kingdom  
525460  United Kingdom  

[406295 rows x 8 columns]

In [37]:
cleaned_df.describe()

Quantity                    InvoiceDate  \
count            406295.00                         406295   
mean                 13.62  2010-07-01 10:18:05.997415424   
min                   1.00            2009-12-01 07:45:00   
25%                   2.00            2010-03-26 14:01:00   
50%                   5.00            2010-07-09 15:48:00   
75%                  12.00            2010-10-14 17:09:00   
max               19152.00            2010-12-09 20:01:00   
std                  97.00                            NaN   

                     Price          Customer ID  
count            406295.00            406295.00  
mean                  2.99             15373.75  
min                   0.03             12346.00  
25%                   1.25             14006.00  
50%                   1.95             15326.00  
75%                   3.75             16814.00  
max                 295.00             18287.00  
std                   4.29              1677.33

In [38]:
cleaned_df.describe(include=[object])

Invoice StockCode                         Description         Country
count   406295    406295                              406295          406295
unique   18966      4006                                4432              37
top     500356    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq       270      3153                                3153          370455

Much better! I'm okay with 3p items! Let's quickly note the number of rows removed before continuing to the next step.

In [39]:
len(cleaned_df) / len(df)

0.7732162805612596

We removed about `23%` of our original data. Cool! Let's move onto feature engineering.

# Feature Engineering

This is where the real fun begins! Now that our data is cleaned, we need to create some extra features that help us to understand our customers buying habits. Here's what we need:

- The number of purchases the customer has made (frequency)
- The amount they have spent overall (monetary value)
- The date of the last purchase they made (last invoice date)

We'll first start by creating a `SalesTotal` helper column and then create an aggregated DataFrame to group our records.

In [40]:
cleaned_df.reset_index(inplace=True, drop=True)

In [41]:
cleaned_df["SalesTotal"] = cleaned_df["Quantity"] * cleaned_df["Price"]
cleaned_df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
406290  538171     22271                 FELTCRAFT DOLL ROSIE         2   
406291  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
406292  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
406293  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
406294  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate                Price          Customer ID  \
0      2009-12-01 07:45:00                 6.95             13085.00   
1      2009-12-01 07:45:00                 6.75             13085.00   
2      2009-12-01 07:45:00                 6.75             13085.00   
3      2009-12-01 07:45:00                 2.10             13085.00   
4      2009-12-01 07:45:00                 1.25             13085.00   
...                    ...                  ...                  ...   
406290 2010-12-09 20:01:00                 2.95             17530.00   
406291 2010-12-09 20:01:00                 3.75             17530.00   
406292 2010-12-09 20:01:00                 3.75             17530.00   
406293 2010-12-09 20:01:00                 3.75             17530.00   
406294 2010-12-09 20:01:00                 1.95             17530.00   

               Country           SalesTotal  
0       United Kingdom                83.40  
1       United Kingdom                81.00  
2       United Kingdom                81.00  
3       United Kingdom               100.80  
4       United Kingdom                30.00  
...                ...                  ...  
406290  United Kingdom                 5.90  
406291  United Kingdom                 3.75  
406292  United Kingdom                 3.75  
406293  United Kingdom                 7.50  
406294  United Kingdom                 3.90  

[406295 rows x 9 columns]

To achieve our desired results, we group the `Customer ID` with some new columns:
- `MonetaryValue` - for the total cost of the customers order
- `Frequency` - the number of unique invoices
- `LastInvoiceDate` - the last invoice (purchase made) generated by that customer

In [42]:
aggregated_df = cleaned_df.groupby(by="Customer ID", as_index=False) \
                .agg(
                    MonetaryValue=("SalesTotal", "sum"), 
                    Frequency=("Invoice", "nunique"),
                    LastInvoiceDate=("InvoiceDate", "max")
                )
aggregated_df

Customer ID        MonetaryValue  Frequency     LastInvoiceDate
0                12346.00               169.36          2 2010-06-28 13:53:00
1                12347.00              1323.32          2 2010-12-07 14:57:00
2                12348.00               221.16          1 2010-09-27 14:59:00
3                12349.00              2221.14          2 2010-10-28 08:23:00
4                12351.00               300.93          1 2010-11-29 15:23:00
...                   ...                  ...        ...                 ...
4280             18283.00               641.77          6 2010-11-22 15:30:00
4281             18284.00               411.68          1 2010-10-04 11:33:00
4282             18285.00               377.00          1 2010-02-17 10:24:00
4283             18286.00              1246.43          2 2010-08-20 11:57:00
4284             18287.00              2295.71          4 2010-11-22 11:51:00

[4285 rows x 4 columns]

Perfect! Here we can see that we have 4,285 unique customers. Let's also add a `Recency` column. For this dataset, we'll use the last invoice date, but normally, we'd use the date of the analysis (e.g., today's date). However, that doesn't make sense given the age of this dataset.

Also, to keep things simple, we'll specify the recency in days. It's easier to work with computationally.

In [43]:
max_invoice_date = aggregated_df["LastInvoiceDate"].max()

aggregated_df["RecencyDays"] = (max_invoice_date - aggregated_df['LastInvoiceDate']).dt.days
aggregated_df

Customer ID        MonetaryValue  Frequency     LastInvoiceDate  \
0                12346.00               169.36          2 2010-06-28 13:53:00   
1                12347.00              1323.32          2 2010-12-07 14:57:00   
2                12348.00               221.16          1 2010-09-27 14:59:00   
3                12349.00              2221.14          2 2010-10-28 08:23:00   
4                12351.00               300.93          1 2010-11-29 15:23:00   
...                   ...                  ...        ...                 ...   
4280             18283.00               641.77          6 2010-11-22 15:30:00   
4281             18284.00               411.68          1 2010-10-04 11:33:00   
4282             18285.00               377.00          1 2010-02-17 10:24:00   
4283             18286.00              1246.43          2 2010-08-20 11:57:00   
4284             18287.00              2295.71          4 2010-11-22 11:51:00   

      RecencyDays  
0             164  
1               2  
2              73  
3              42  
4              10  
...           ...  
4280           17  
4281           66  
4282          295  
4283          111  
4284           17  

[4285 rows x 5 columns]

# Visual Exploration

Okay, now let's start visualising our data's distributions and identify any outliers.